pkg/yurthub/healthchecker/health_checker.go

```diff
@@ -21,114 +21,368 @@ import (
+	holderIdentity          string
+	leaseDurationSeconds    int32
+	clock                   clock.Clock
+	failedRetry             int
+	tp                      transport.Interface
+	sw                      cachemanager.StorageWrapper
+	healthyThreshold        int
+	heartbeatTimeoutSeconds time.Duration
+
+	remoteServers     []*url.URL
+	remoteServerIndex int
+
+	stopCh <-chan struct{}
 }
 
 // NewHealthChecker create an HealthChecker for servers
-func NewHealthChecker(remoteServers []*url.URL, tp transport.Interface, failedRetry, healthyThreshold int, stopCh <-chan struct{}) (HealthChecker, error) {
+func NewHealthChecker(remoteServers []*url.URL, tp transport.Interface, sw cachemanager.StorageWrapper, failedRetry, healthyThreshold, heartBeatTimeout int, stopCh <-chan struct{}) (HealthChecker, error) {
 	if len(remoteServers) == 0 {
 		return nil, fmt.Errorf("no remote servers")
 	}
 
 	hcm := &healthCheckerManager{
-		checkers: make(map[string]*checker),
+		checkers:                make(map[string]*checker),
+		remoteServers:           remoteServers,
+		remoteServerIndex:       0,
+		failedRetry:             failedRetry,
+		holderIdentity:          os.Getenv("NODE_NAME"),
+		tp:                      tp,
+		sw:                      sw,
+		healthyThreshold:        healthyThreshold,
+		heartbeatTimeoutSeconds: time.Duration(heartBeatTimeout) * time.Second,
+		leaseDurationSeconds:    defaultLeaseDurationSeconds,
+		clock:                   clock.RealClock{},
+		stopCh:                  stopCh,
 	}
 
-	for _, server := range remoteServers {
-		checker, err := newChecker(server, tp, failedRetry, healthyThreshold, stopCh)
-		if err != nil {
-			klog.Errorf("new health checker for %s err, %v", server.String(), err)
-			return nil, err
+	return hcm, nil
+}
+
+func (hcm *healthCheckerManager) Run() {
+
+	for _, server := range hcm.remoteServers {
+		if _, ok := hcm.checkers[server.String()]; !ok {
+			checker, err := hcm.newChecker(server)
+			if err != nil {
+				klog.Errorf("new health checker for %s err, %v", server.String(), err)
+				return
+			}
+			hcm.checkers[server.String()] = checker
 		}
-		hcm.checkers[server.String()] = checker
 	}
 
-	return hcm, nil
+	if err := hcm.storageLease(hcm.latestLease); err != nil {
+		klog.Errorf("storage lease fail: %v", err)
+	}
+
+	go hcm.healthzCheckLoop(hcm.stopCh)
+
+	return
 }
 
-// IsHealthy returns the healthy stats of specified server
-func (hcm *healthCheckerManager) IsHealthy(server *url.URL) bool {
-	hcm.RLock()
-	defer hcm.RUnlock()
-	if checker, ok := hcm.checkers[server.String()]; ok {
-		return checker.isHealthy()
+func (hcm *healthCheckerManager) healthzCheckLoop(stopCh <-chan struct{}) {
+	intervalTicker := time.NewTicker(heartbeatFrequency)
+	defer intervalTicker.Stop()
+
+	for {
+		select {
+		case <-stopCh:
+			klog.Infof("exit normally in health check loop.")
+			return
+		case <-intervalTicker.C:
+			hcm.sync()
+		}
 	}
+}
 
-	return false
+func (hcm *healthCheckerManager) sync() {
```

Comment:
Please add comments to explain the business of checking healthy status of server by lease?

Reply:
Ensure that the node heartbeat can be reported when there is a healthy remote server